logstash-core/src/main/java/org/logstash/config/ir/compiler/Dataset.java

```diff
@@ -0,0 +1,438 @@
+     */
+    void clear();
+
+    /**
+     * Root {@link Dataset}s at the beginning of the execution tree that simply pass through
+     * the given set of {@link JrubyEventExtLibrary.RubyEvent} and have no state.
+     */
+    Collection<Dataset> ROOT_DATASETS = Collections.singleton(
+        new Dataset() {
+            @Override
+            public Collection<JrubyEventExtLibrary.RubyEvent> compute(
+                final RubyIntegration.Batch batch, final boolean flush, final boolean shutdown,
+                final RubyHash options) {
+                return batch.collect();
+            }
+
+            @Override
+            public void clear() {
+            }
+        }
+    );
+
+    /**
+     * <p>{@link Dataset} that contains all {@link JrubyEventExtLibrary.RubyEvent} of all of
+     * its dependencies and is assumed to be at the end of an execution.</p>
+     * This dataset does not require an explicit call to {@link Dataset#clear()} since it will
+     * automatically {@code clear} all of its parents.
+     */
+    final class TerminalDataset implements Dataset {
+
+        /**
+         * Empty {@link Collection} returned by this class's
+         * {@link Dataset#compute(RubyIntegration.Batch, boolean, boolean, RubyHash)} implementation.
+         */
+        private static final Collection<JrubyEventExtLibrary.RubyEvent> EMPTY_RETURN =
+            Collections.emptyList();
+
+        /**
+         * Trivial {@link Dataset} that simply returns an empty collection of elements.
+         */
+        private static final Dataset TRIVIAL = new Dataset() {
+
+            @Override
+            public Collection<JrubyEventExtLibrary.RubyEvent> compute(
+                final RubyIntegration.Batch batch, final boolean flush, final boolean shutdown,
+                final RubyHash options) {
+                return EMPTY_RETURN;
+            }
+
+            @Override
+            public void clear() {
+            }
+        };
+
+        private final Collection<Dataset> parents;
+
+        /**
+         * <p>Builds a terminal {@link Dataset} from the given parent {@link Dataset}s.</p>
+         * <p>If the given set of parent {@link Dataset} is empty the sum is defined as the
+         * trivial dataset that does not invoke any computation whatsoever.</p>
+         * The return of a call to
+         * {@link Dataset#compute(RubyIntegration.Batch, boolean, boolean, RubyHash)} is always
+         * {@link Collections#emptyList()}.
+         * @param parents Parent {@link Dataset} to sum and terminate
+         * @return Dataset representing the sum of given parent {@link Dataset}
+         */
+        public static Dataset from(final Collection<Dataset> parents, final boolean debug) {
+            final int count = parents.size();
+            final Dataset result;
+            if (count > 0) {
+                if (debug) {
+                    result = new Dataset.TerminalDataset.TerminalDebugDataset(parents);
```

Comment:
Sometimes we pass debug as a flag (here), and sometimes the "debug" part is implemented as a class or a function by a special name w/ "Debug" in the name.

Why choose `from(..., debug)` here instead of, similar to elsewhere in this PR, having two methods:

```java
public static Dataset from(final Collection<Dataset> parents) { ... }

public static Dataset fromDebug(final Collection<Dataset> parents) { ... }
```

I'm not strictly in favor of one (method/classname indicating "Debug", vs a boolean parameter), but just highlighting some inconsistency.